blockchain/chainquery.go

```diff
@@ -154,3 +154,80 @@ func (b *BlockChain) BestInvalidHeader() chainhash.Hash {
 	b.index.RUnlock()
 	return hash
 }
+
+// NextNeededBlocks returns hashes for the next blocks after the current best
+// chain tip that are needed to make progress towards the current best known
+// header skipping any blocks that are already known or in the provided map of
+// blocks to exclude.  Typically the caller would want to exclude all blocks
+// that have outstanding requests.
+//
+// The maximum number of results is limited to the provided value or the maximum
+// allowed by the internal lookahead buffer in the case the requested number of
+// max results exceeds that value.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) NextNeededBlocks(maxResults uint8, exclude map[chainhash.Hash]struct{}) []*chainhash.Hash {
+	// Nothing to do when no results are requested.
+	if maxResults == 0 {
+		return nil
+	}
+
+	// No blocks are needed from branches that do not lead to the current best
+	// header.
+	tip := b.bestChain.Tip()
+	b.index.RLock()
+	bestHeader := b.index.bestHeader
+	if bestHeader.Ancestor(tip.height) != tip {
+		b.index.RUnlock()
+		return nil
+	}
```

Comment:
This doesn't seem to be the case, since while doing a simnet test, reorgs are working just fine so I'm misunderstanding something, just not clear yet what exactly.

Reply:
You're right.  This is actually some code from the multi-download that is under development too that snuck in when I was splitting this bit off.  Nice catch!  I'll get this updated here in a bit.

The reason reorgs are still working for you as expected is because there are currently two paths.  One is the catchup path, which is this one.  The other one is the immediate download path that happens once you are current for better throughput, mainly for votes at the current time.

Reply:
Updated.